src/main/java/Duke/Storage.java

```diff
@@ -0,0 +1,199 @@
+
+package Duke;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The Storage class handles the saving and loading of tasks from a file.
+ */
+public class Storage {
+    private final String filePath;
+
+    /**
+     * Constructs a Storage instance with the specified file path.
+     *
+     * @param filePath The path to the data file.
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+    /**
+     * Saves the tasks to the data file.
+     *
+     * @param tasks The TaskList containing tasks to be saved.
+     * @throws IOException If an error occurs while saving the file.
+     */
+    public void saveTasksToFile(TaskList tasks) throws IOException {
+        Path filePath = Paths.get(this.filePath);
+
+        // Create the parent directory if it doesn't exist
+        if (!Files.exists(filePath.getParent())) {
+            Files.createDirectories(filePath.getParent());
+        }
+
+        List<String> encodedTasks = TaskListEncoder.encode(tasks.getTasks());
+        // Write tasks to the file
+        Files.write(filePath, encodedTasks);
+    }
+
+    /**
+     * Loads tasks from the data file.
+     *
+     * @return The loaded TaskList.
+     * @throws IOException              If an error occurs while reading the file.
+     * @throws DukeDataCorruptedException If the data file is corrupted.
+     */
+    public TaskList loadTasksFromFile() throws IOException, DukeDataCorruptedException {
+        Path filePath = Paths.get(this.filePath);
+
+        TaskList tasks;
+
+        // Check if the file exists before attempting to load
+        if (Files.exists(filePath)) {
+            try {
+                // Read tasks from the file
+                List<String> lines = Files.readAllLines(filePath);
+                tasks = new TaskList(TaskListDecoder.decode(lines));
+            } catch (FileNotFoundException e) {
+                // If the file is not found, throw a more specific exception
+                throw new FileNotFoundException("Data file not found");
+            } catch (IOException e) {
+                throw new IOException("Error reading tasks from the file.", e);
+            } catch (DukeDataCorruptedException e) {
+                // Rethrow other exceptions, including DukeDataCorruptedException
+                throw e;
+            }
+        } else {
+            tasks = new TaskList(); // Initialize with an empty list if the file doesn't exist yet
+        }
+
+        return tasks;
+    }
+
+    /**
+     * Ensures that the data file exists; if not, it creates the file.
+     *
+     * @throws IOException If an error occurs while creating the file.
+     */
+    public void ensureDataFileExists() throws IOException {
```

Comment:
Perhaps a better method name would be createDateFile as it informs the client that the method takes an action that alters the state of the application. createDataFileIfNotExisting is even more descriptive